### PR TITLE
A setup for Clippy

### PR DESCRIPTION
**Pull request by ChristianBeilschmidt**
This PR is a suggestion for incorporating some of the "pedantic" lints.
Moreover, it changes the CI to fail on missed issues.
This prevents us from saying: "Please check all linter issues".

### CODE REVIEW

File: datatypes/src/lib.rs [hunk 1]
```diff
@@ -1,3 +1,22 @@
...
+    clippy::cast_precision_loss,
+    clippy::cast_sign_loss,
+    clippy::default_trait_access,
+    clippy::module_name_repetitions,
+    clippy::must_use_candidate,
+    clippy::non_ascii_literal,
+    clippy::similar_names,
+    clippy::single_match_else,
+    clippy::type_repetition_in_bounds,
+    clippy::wildcard_imports
```
Comment (michaelmattig) [inline datatypes/src/lib.rs:17]: `warning: unknown clippy lint: clippy::wildcard_imports`

Comment (michaelmattig) [reply datatypes/src/lib.rs:17]: I would also allow `missing_errors_doc` as I think this will often be straightforward and result in boiler doc

Comment (michaelmattig) [reply datatypes/src/lib.rs:17]: `map_flatten` and `filter_map` could also be allowed

Comment (michaelmattig) [reply datatypes/src/lib.rs:17]: it's fine for me to fail CI when there are linter errors

**Comment by ChristianBeilschmidt** [comment]
I allowed `missing_errors_doc` and vote for trying out `map_flatten` and `filter_map` until they annoy us.

**Pull request merged**